libretro-common/include/libretro.h

```diff
@@ -3030,6 +3052,89 @@ struct retro_disk_control_ext_callback
    retro_get_image_label_t get_image_label;     /* Optional - may be NULL */
 };
 
+/* Callbacks for RETRO_ENVIRONMENT_SET_NETPACKET_INTERFACE.
+ * A core can set it if sending and receiving custom network packets
+ * during a multiplayer session is desired.
+ */
+
+/* Used by the core to send a packet to one or more connected players.
+ * A single packet sent via this interface can contain up to 64kb of data.
+ *
+ * If the ready callback has indicated the local player to be the host:
+ *  - The broadcast flag can be set to true to send to multiple connected clients
+ *  - On a broadcast, the client_id argument indicates 1 client NOT to send the packet to
+ *  - Otherwise, the client_id argument indicates a single client to send the packet to
+ * If the local player is a client connected to a host:
+ *  - The broadcast flag is ignored
+ *  - The client_id argument must be set to 0
+ *
+ * This function is not guaranteed to be thread-safe and must be called during
+ * retro_run or any of the netpacket callbacks passed with this interface.
+ */
+typedef void (RETRO_CALLCONV *retro_netpacket_send_t)(const void* buf, size_t len, uint16_t client_id, bool broadcast);
```

Comment:
I can think of a use case for all clients being able to talk to each other (rather than just the host). [This thread on Discord has the details](https://discord.com/channels/184109094070779904/876520593636335646/1120811822182834226), but in a nutshell:

- Handheld consoles like the Nintendo DS or Game Boy support local multiplayer via link cables or custom wireless protocols.
- These protocols usually have extremely tight latency requirements that can't be met over the Internet.
- The best general solution (currently adopted by [at least one Game Boy core](https://github.com/libretro/tgbdual-liretro)) is for all clients to emulate all consoles, and synchronize the results via rollback.
- Not to be confused with true Internet connectivity, like the Wi-fi services offered by the Wii and DS. The original hardware had to account for Wi-fi latency, so emulators can support it without rollback tricks.

Given this, a practical use case for P2P communication would be to send savestates of all consoles to all other participants.

Reply:
Yeah, I think it would be feasible to allow clients talking to clients via this directly.
In RetroArch netplay this always will have to relayed via the host (i.e. CLIENTA -> HOST -> CLIENTB) but perhaps in another frontend it could be a different kind of network setup.
Currently in the implementation I did in my DOS core I have such a packet relaying done inside the core, which is aware if it acts as the host and it does the packet relaying there. But we could move that into the frontend and change the API to allow a client talking to another client. I'll see if that can be done.

As for using this in a tight latency requirements, an easy first try might be to do lockstep without rollback. This new interface has this specified:
```
* When 2 or more players are connected and this interface has been
* set, time manipulation features (pausing, slow motion, fast forward,
* rewinding, save state loading, etc.) are disabled to not interrupt
* communication.
```
So when this is set up and cores can talk to each other, the frontend will not mess with timing. Thus a core could take over and only advance emulation when the data over the emulated local wi-fi or link cable has arrived.
Once that works, rollback could be implemented on top of that. Don't lockstep anymore but roll back when wifi/cable data from the past arrives.

As for sending whole savestates over this, the limitation of 64kb of data per packet might be annoying. Though splitting it up into multiple packets wouldn't be too hard. The 64kb limit I chose pretty arbitrarily, we could even remove it and have the frontend realloc the buffer it uses for this once a core wants to send a bigger packet at once.

Reply:
The new commit c0f85f8 now adds the option for clients to send packets directly to other clients.

This means all peers are equal in regards to what `retro_netpacket_send_t` can do. Anyone can send a packet to any other peer as well as send broadcasts to everyone connected.

Because in RetroArch netplay clients don't connect to each other (they only connect to the host) it is implemented by having the host relay packets as needed. If the core on client A sends a packet addressed to client B, the netplay host will relay that packet but it won't call `retro_netpacket_receive_t` locally. Another frontend could have a mesh connection where clients can directly talk to each other, resulting in the core experiencing the same behavior.